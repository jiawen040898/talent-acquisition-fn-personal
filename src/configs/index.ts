export * from './auth.config';
export * from './aws.config';
export * from './skill-extract-api.config';
export * from './unleash.config';
