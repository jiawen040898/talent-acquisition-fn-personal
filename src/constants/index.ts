export * from './action-history.enum';
export * from './auth-error.enum';
export * from './cache-store.enum';
export * from './candidate.event-type.enum';
export * from './channel.enum';
export * from './contact-type.enum';
export * from './ds-lookup-api-errors.enum';
export * from './email-event.enum';
export * from './email-type.enum';
export * from './error-msg.enum';
export * from './errors';
export * from './feature-toggle-name.enum';
export * from './fit-score-recipe.enum';
export * from './identity-event.enum';
export * from './job.enum';
export * from './job-application-score-type.enum';
export * from './job-application-skill.enum';
export * from './job-application-status.enum';
export * from './person-score.enum';
export * from './person-score-event.enum';
export * from './psychology-api-errors.enum';
export * from './questionnaire-framework.enum';
export * from './s3-errors.enum';
export * from './screening-answer-criteria.enum';
export * from './skill-extract-api-errors.enum';
export * from './talent-acquisition-event.enum';
export * from './talent-management-event.enum';
export * from './timeline.enum';
export * from './unleash-errors.enum';
export * from './user.enum';
