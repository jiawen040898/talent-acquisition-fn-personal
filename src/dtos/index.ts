export * from './audit-data.dto';
export * from './candidate-event.dto';
export * from './create-job.dto';
export * from './fit-score-recipe.dto';
export * from './geolocation.dto';
export * from './identity-update-user-event.dto';
export * from './job-application.dto';
export * from './job-application-action-history.dto';
export * from './job-application-attachment.dto';
export * from './job-application-career.dto';
export * from './job-application-contact.dto';
export * from './job-application-education.dto';
export * from './job-application-questionnaire.dto';
export * from './job-application-resume.dto';
export * from './job-application-score.dto';
export * from './job-application-screening-answer.dto';
export * from './person-score-event.dto';
export * from './place.dto';
export * from './talent-acquisition-event.dto';
export * from './talent-management-event.dto';
