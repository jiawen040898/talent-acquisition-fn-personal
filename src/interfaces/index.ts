export * from './api.interface';
export * from './api-response.interface';
export * from './aws.interface';
export * from './candidate.interface';
export * from './cognitive-score.interface';
export * from './email.interface';
export * from './event-model.interface';
export * from './job-application-score.interface';
export * from './pairwise-score.interface';
export * from './personality.interface';
export * from './personality-score.interface';
export * from './skill-extract-api.interface';
export * from './user.interface';
