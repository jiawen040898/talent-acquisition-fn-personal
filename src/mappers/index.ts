export * from './job-application-mapper';
export * from './job-application-score-mapper';
