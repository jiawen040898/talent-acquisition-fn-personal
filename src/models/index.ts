export * from './audit-data.entity';
export * from './entities';
export * from './job.entity';
export * from './job-application.entity';
export * from './job-application-action-history.entity';
export * from './job-application-attachment.entity';
export * from './job-application-career.entity';
export * from './job-application-contact.entity';
export * from './job-application-education.entity';
export * from './job-application-label.entity';
export * from './job-application-questionnaire.entity';
export * from './job-application-resume.entity';
export * from './job-application-score.entity';
export * from './job-application-screening-answer.entity';
export * from './job-distribution-score.entity';
export * from './job-questionnaire.entity';
export * from './job-screening-question.entity';
export * from './job-user-access.entity';
