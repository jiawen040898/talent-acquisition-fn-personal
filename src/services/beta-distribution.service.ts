export const EXPONENTIAL_SMOOTHING_WEIGHT = 0.025; // Gamma

const smooth = (alpha: number, beta: number, gamma: number): number => {
    return alpha + gamma * (beta - alpha);
};

const getGammaForUniformSmoothing = (size: number): number => {
    return 1 / size;
};

const getNewMean = (
    oldMean: number,
    value: number,
    weight = EXPONENTIAL_SMOOTHING_WEIGHT,
): number => {
    return smooth(oldMean, value, weight);
};

const getNewVariance = (
    oldVariance: number,
    oldMean: number,
    newMean: number,
    value: number,
    weight = EXPONENTIAL_SMOOTHING_WEIGHT,
): number => {
    return smooth(oldVariance, (value - oldMean) * (value - newMean), weight);
};

const getAlphaAndBeta = (
    mean: number,
    variance: number,
    size: number,
): {
    alpha: number;
    beta: number;
} => {
    if (size === 1) {
        return {
            alpha: 1,
            beta: 1,
        };
    }

    const alpha = mean * ((mean * (1 - mean)) / variance - 1);
    const beta = (1 - mean) * ((mean * (1 - mean)) / variance - 1);
    return { alpha, beta };
};

export const BetaDistributionService = {
    getGammaForUniformSmoothing,
    getNewMean,
    getNewVariance,
    getAlphaAndBeta,
};
