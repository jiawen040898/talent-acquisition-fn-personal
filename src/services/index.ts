export * from './beta-distribution.service';
export * from './cognitive-score.service';
export * from './ds-lookup.service';
export * from './feature-toggle.service';
export * from './fit-score.service';
export * from './job-application.service';
export * from './job-application-score.service';
export * from './job-application-skill-extraction.service';
export * from './job-distribution-score.service';
export * from './job-user-access.service';
export * from './pairwise-score.service';
export * from './personality-score.service';
export * from './psychology.service';
export * from './publisher.service';
export * from './s3.service';
export * from './skill-extract-api.service';
