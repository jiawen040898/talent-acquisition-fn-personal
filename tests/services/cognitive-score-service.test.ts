import { CognitiveScoreService } from '@pulsifi/services';
import {
    testJobApplicationBuilder,
    testJobApplicationScoreBuilder,
    testJobBuilder,
} from '@pulsifi/tests/builders';
import {
    completedQuestionnaires,
    mockRoleFitRecipe,
    reasoningNumericQuestionnaire,
} from '@pulsifi/tests/fixtures';
import {
    logicalPartialScoreOutput,
    numericPartialScoreOutput,
    reasoningNumericalScore,
    verbalPartialScoreOutput,
} from '@pulsifi/tests/fixtures/job-application-score/test-data';
import { getTestDataSource } from '@pulsifi/tests/setup';
import { DataSource } from 'typeorm';

import { getDataSource, setDataSource } from '../../src/database';

describe('CognitiveScoreService', () => {
    let dataSource: DataSource;
    const jobApplicationId = '00000000-0000-0000-0001-000000000001';
    const jobApplicationId2 = '00000000-0000-0000-0001-000000000002';
    const jobApplicationId3 = '00000000-0000-0000-0001-000000000003';

    beforeAll(async () => {
        setDataSource(await getTestDataSource());
        dataSource = await getDataSource();

        const job = testJobBuilder.build({
            id: '00000000-0000-0000-0001-100000000001',
        });
        const jobApplication = testJobApplicationBuilder.build({
            id: jobApplicationId,
            job_id: job.id,
        });
        const jobApplication2 = testJobApplicationBuilder.build({
            id: jobApplicationId2,
            job_id: job.id,
        });
        const jobApplication3 = testJobApplicationBuilder.build({
            id: jobApplicationId3,
            job_id: job.id,
        });
        const jobApplicationScore = testJobApplicationScoreBuilder.build({
            ...reasoningNumericalScore,
            id: undefined,
            job_application_id: jobApplication.id,
        });
        const jobApplicationScore2 = testJobApplicationScoreBuilder.build({
            ...reasoningNumericalScore,
            id: undefined,
            job_application_id: jobApplication3.id,
        });

        await dataSource.getRepository('Job').save(job);
        await dataSource
            .getRepository('JobApplication')
            .save([jobApplication, jobApplication2, jobApplication3]);
        await dataSource
            .getRepository('JobApplicationScore')
            .save([jobApplicationScore, jobApplicationScore2]);
    });

    describe('getCognitiveScores', () => {
        it('should return correct output', async () => {
            // Act
            const result = await CognitiveScoreService.getCognitiveScores(
                completedQuestionnaires,
                jobApplicationId,
                5,
                mockRoleFitRecipe.id,
            );

            // Assert
            expect(result).toMatchSnapshot();
        });

        it('should still be able to calculate average score when not all cognitive are calculated', async () => {
            // Act
            const result = await CognitiveScoreService.getCognitiveScores(
                [reasoningNumericQuestionnaire],
                '00000000-0000-0000-0000-000000000002',
                5,
                mockRoleFitRecipe.id,
            );

            // Assert
            expect(result).toMatchSnapshot();
        });
    });

    describe('processReasoningAverage', () => {
        it('should return correct output', async () => {
            // Act
            const result = await CognitiveScoreService.processReasoningAverage(
                [
                    verbalPartialScoreOutput,
                    logicalPartialScoreOutput,
                    numericPartialScoreOutput,
                ],
                jobApplicationId,
                5,
            );

            // Assert
            expect(result).toMatchSnapshot();
        });

        it('should return correct output with incomplete average', async () => {
            // Act
            const result = await CognitiveScoreService.processReasoningAverage(
                [numericPartialScoreOutput],
                jobApplicationId,
                5,
            );

            // Assert
            expect(result).toMatchSnapshot();
        });

        it('should return correct output with complete score combining with db', async () => {
            // Act
            const result = await CognitiveScoreService.processReasoningAverage(
                [logicalPartialScoreOutput, verbalPartialScoreOutput],
                jobApplicationId3,
                5,
            );

            // Assert
            expect(result).toMatchSnapshot();
        });
    });

    describe('processCognitiveScore', () => {
        it('should return correct output', () => {
            // Act
            const result = CognitiveScoreService.processCognitiveScore(
                reasoningNumericQuestionnaire,
                jobApplicationId,
                5,
            );

            // Assert
            expect(result).toMatchSnapshot();
        });
    });
});
