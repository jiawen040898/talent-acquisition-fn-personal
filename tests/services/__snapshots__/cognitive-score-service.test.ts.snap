// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CognitiveScoreService getCognitiveScores should return correct output 1`] = `
[
  {
    "created_by": 5,
    "job_application_id": "00000000-0000-0000-0001-000000000001",
    "score": 73.2324325,
    "score_dimension": 2,
    "score_outcome": {
      "cognitive_result": {
        "domain_alias": "verbal",
        "domain_percentile": null,
        "domain_score": 0.732324325294115,
        "ingredient_weightage": null,
      },
    },
    "score_recipe_id": "106e0b1d-02c4-4321-9dac-9768f84f8bc1",
    "score_type": "reasoning_verbal",
    "updated_by": 5,
  },
  {
    "created_by": 5,
    "job_application_id": "00000000-0000-0000-0001-000000000001",
    "score": 77.8327824,
    "score_dimension": 2,
    "score_outcome": {
      "cognitive_result": {
        "domain_alias": "logical",
        "domain_percentile": null,
        "domain_score": 0.7783278238,
        "ingredient_weightage": null,
      },
    },
    "score_recipe_id": "106e0b1d-02c4-4321-9dac-9768f84f8bc1",
    "score_type": "reasoning_logical",
    "updated_by": 5,
  },
  {
    "created_by": 5,
    "job_application_id": "00000000-0000-0000-0001-000000000001",
    "score": 33.7397,
    "score_dimension": 1,
    "score_outcome": {
      "cognitive_result": {
        "domain_alias": "numeric",
        "domain_percentile": null,
        "domain_score": 0.337397,
        "ingredient_weightage": null,
      },
    },
    "score_recipe_id": "106e0b1d-02c4-4321-9dac-9768f84f8bc1",
    "score_type": "reasoning_numeric",
    "updated_by": 5,
  },
  {
    "created_by": 5,
    "job_application_id": "00000000-0000-0000-0001-000000000001",
    "score": 61.6016383,
    "score_dimension": 1,
    "score_outcome": {
      "cognitive_result": {
        "domain_alias": "reasoning_average",
        "domain_percentile": null,
        "domain_score": 0.616016383,
        "ingredient_weightage": null,
      },
    },
    "score_recipe_id": "106e0b1d-02c4-4321-9dac-9768f84f8bc1",
    "score_type": "reasoning_average",
    "updated_by": 5,
  },
]
`;

exports[`CognitiveScoreService getCognitiveScores should still be able to calculate average score when not all cognitive are calculated 1`] = `
[
  {
    "created_by": 5,
    "job_application_id": "00000000-0000-0000-0000-000000000002",
    "score": 33.7397,
    "score_dimension": 1,
    "score_outcome": {
      "cognitive_result": {
        "domain_alias": "numeric",
        "domain_percentile": null,
        "domain_score": 0.337397,
        "ingredient_weightage": null,
      },
    },
    "score_recipe_id": "106e0b1d-02c4-4321-9dac-9768f84f8bc1",
    "score_type": "reasoning_numeric",
    "updated_by": 5,
  },
  {
    "created_by": 5,
    "job_application_id": "00000000-0000-0000-0000-000000000002",
    "score": 33.7397,
    "score_dimension": 1,
    "score_outcome": {
      "cognitive_result": {
        "domain_alias": "reasoning_average",
        "domain_percentile": null,
        "domain_score": 0.337397,
        "ingredient_weightage": null,
      },
    },
    "score_recipe_id": "106e0b1d-02c4-4321-9dac-9768f84f8bc1",
    "score_type": "reasoning_average",
    "updated_by": 5,
  },
]
`;

exports[`CognitiveScoreService processCognitiveScore should return correct output 1`] = `
{
  "created_by": 5,
  "job_application_id": "00000000-0000-0000-0001-000000000001",
  "score": 33.7397,
  "score_dimension": 1,
  "score_outcome": {
    "cognitive_result": {
      "domain_alias": "numeric",
      "domain_percentile": null,
      "domain_score": 0.337397,
      "ingredient_weightage": null,
    },
  },
  "score_recipe_id": undefined,
  "score_type": "reasoning_numeric",
  "updated_by": 5,
}
`;

exports[`CognitiveScoreService processReasoningAverage should return correct output 1`] = `
{
  "created_by": 5,
  "job_application_id": "00000000-0000-0000-0001-000000000001",
  "score": 61.601638,
  "score_dimension": 1,
  "score_outcome": {
    "cognitive_result": {
      "domain_alias": "reasoning_average",
      "domain_percentile": null,
      "domain_score": 0.6160163799999999,
      "ingredient_weightage": null,
    },
  },
  "score_recipe_id": undefined,
  "score_type": "reasoning_average",
  "updated_by": 5,
}
`;

exports[`CognitiveScoreService processReasoningAverage should return correct output with complete score combining with db 1`] = `
{
  "created_by": 5,
  "job_application_id": "00000000-0000-0000-0001-000000000003",
  "score": 78.5342803,
  "score_dimension": 2,
  "score_outcome": {
    "cognitive_result": {
      "domain_alias": "reasoning_average",
      "domain_percentile": null,
      "domain_score": 0.7853428027999999,
      "ingredient_weightage": null,
    },
  },
  "score_recipe_id": undefined,
  "score_type": "reasoning_average",
  "updated_by": 5,
}
`;

exports[`CognitiveScoreService processReasoningAverage should return correct output with incomplete average 1`] = `
{
  "created_by": 5,
  "job_application_id": "00000000-0000-0000-0001-000000000001",
  "score": 33.7397,
  "score_dimension": 1,
  "score_outcome": {
    "cognitive_result": {
      "domain_alias": "reasoning_average",
      "domain_percentile": null,
      "domain_score": 0.337397,
      "ingredient_weightage": null,
    },
  },
  "score_recipe_id": undefined,
  "score_type": "reasoning_average",
  "updated_by": 5,
}
`;
